src/burn/engine/apply.cpp

```diff
@@ -739,6 +744,9 @@ extern "C" HRESULT ApplyExecute(
                 IgnoreRollbackError(hrRollback, "Failed rollback actions");
             }
 
+            hrRollback = DoRestoreRelatedBundleActions(pEngineState, &context, pRestart);
```

Comment:
3. Identifying which upgrade bundle was responsible for each MSI that was uninstalled is not always possible. Another wrinkle here is that Exe packages could also be breaking things that the older bundles needs to repair.